swarm/storage/hasherstore.go

```diff
@@ -0,0 +1,223 @@
+func (h *hasherStore) Get(ref Reference) (ChunkData, error) {
+	if h.store == nil {
+		return nil, errors.New("Can not get ref from HasherStore with nil ChunkStore")
+	}
+	key, encryptionKey, err := parseReference(ref)
+	if err != nil {
+		return nil, err
+	}
+	toDecrypt := (encryptionKey != nil)
+
+	chunk, err := h.store.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	chunkData := chunk.SData
+	if toDecrypt {
+		var err error
+		chunkData, err = h.decryptChunkData(chunkData, encryptionKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return chunkData, nil
+}
+
+// The Close() indicates that no more chunks will be put with the hasherStore, so the Wait
+// function can return when all the previously put chunks has been stored.
+func (h *hasherStore) Close() {
+	close(h.closed)
+}
+
+// Wait() function returns when
+//    1) the Close() function has been called and
+//    2) all the chunks which has been Put has been stored
+func (h *hasherStore) Wait() {
+	<-h.closed
+	h.wg.Wait()
+}
+
+func (h *hasherStore) createHash(chunkData ChunkData) Key {
+	hasher := h.hashFunc()
+	hasher.ResetWithLength(chunkData[:8]) // 8 bytes of length
+	hasher.Write(chunkData[8:])           // minus 8 []byte length
+	return hasher.Sum(nil)
+}
+
+func (h *hasherStore) createChunk(chunkData ChunkData, chunkSize int64) *Chunk {
+	hash := h.createHash(chunkData)
+	chunk := NewChunk(hash, nil)
+	chunk.SData = chunkData
+	chunk.Size = chunkSize
+
+	return chunk
+}
+
+func (p *hasherStore) encryptChunkData(chunkData ChunkData) (ChunkData, encryption.Key, error) {
+	if len(chunkData) < 8 {
+		return nil, nil, fmt.Errorf("Invalid ChunkData, min length 8 got %v", len(chunkData))
+	}
+
+	encryptionKey, err := encryption.GenerateRandomKey()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c := make(ChunkData, len(chunkData))
+	encryptedSpan, err := p.chunkEncryption.spanEncryption.Encrypt(chunkData[:8], encryptionKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	encryptedData, err := p.chunkEncryption.dataEncryption.Encrypt(chunkData[8:], encryptionKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	copy(c[:8], encryptedSpan)
+	copy(c[8:], encryptedData)
+	return c, encryptionKey, nil
+}
+
+func (p *hasherStore) decryptChunkData(chunkData ChunkData, encryptionKey encryption.Key) (ChunkData, error) {
+	if len(chunkData) < 8 {
+		return nil, fmt.Errorf("Invalid ChunkData, min length 8 got %v", len(chunkData))
+	}
+
+	c := make(ChunkData, len(chunkData))
+	decryptedSpan, err := p.chunkEncryption.spanEncryption.Decrypt(chunkData[:8], encryptionKey)
+	if err != nil {
+		return nil, err
+	}
+	decryptedData, err := p.chunkEncryption.dataEncryption.Decrypt(chunkData[8:], encryptionKey)
+	if err != nil {
+		return nil, err
+	}
+	copy(c[:8], decryptedSpan)
+	copy(c[8:], decryptedData)
+	return c, nil
+}
+
+func (h *hasherStore) RefSize() int64 {
+	return h.refSize
+}
+
+func (h *hasherStore) storeChunk(chunk *Chunk) {
+	// need to do this parallelly
+	if h.store != nil {
```

Comment:
Why are we checking for `h.store`? Shouldn't we just assume it is there?

Reply:
There is one place where we use it with nil, because we only need it for hashing: https://github.com/ethersphere/go-ethereum/blob/swarm-network-rewrite-encryption/cmd/swarm/hash.go#L41
This cmd line tool just gets the hash without storing data. So I think it makes sense to have a nil store if you only need the "hasher" functionality, but not the "store".